### ISSUE TITLE
feature: support domain-level guardrails with domain-specific model training

### ISSUE THREAD

**Issue by Xunzhuo** [vllm-project/semantic-router#1139]
## Description

Support domain-level guardrail configuration with domain-specific model training. This feature will enable training specialized guardrail models (jailbreak detection, PII detection) for specific domains/categories, allowing different security policies and detection thresholds based on the domain context.

Currently, the system supports:
- Global guardrail configuration with per-decision threshold overrides
- Single unified guardrail model for all domains
- Domain classification using MMLU categories

This enhancement will add:
- **Domain-specific guardrail model training** - Train separate guardrail models for different domains (e.g., healthcare, finance, legal, technical)
- **Per-domain model selection** - Route requests to domain-specific guardrail models based on classified domain
- **Domain-aware training pipelines** - Extend existing LoRA training infrastructure to support domain-specific datasets
- **Flexible configuration** - Allow mixing global and domain-specific guardrail models

### Current Architecture

The system already has strong foundations:

1. **Domain Classification** (`models/mom-domain-classifier`)
   - BERT-based classifier with LoRA adapters
   - Supports 14 MMLU categories
   - Training pipeline: `src/training/training_lora/classifier_model_fine_tuning_lora/ft_linear_lora.py`

2. **Guardrail Models**
   - Jailbreak detection: `models/mom-jailbreak-classifier` (ModernBERT-based, 2 classes)
   - PII detection: `models/mom-pii-classifier` (BERT-based with LoRA, 35 types)
   - Training pipelines in `src/training/training_lora/prompt_guard_fine_tuning_lora/`

3. **Per-Decision Configuration** (already supported)
   - Different thresholds per domain via `decisions[].plugins[].configuration.threshold`
   - Example: `config/prompt-guard/jailbreak_domain.yaml`

### Proposed Enhancement

Enable training and deploying domain-specific guardrail models:

```yaml
# Example configuration
classifier:
  category_model:
    model_id: "models/mom-domain-classifier"
  
  # Domain-specific guardrail models
  domain_guardrails:
    healthcare:
      jailbreak_model: "models/mom-jailbreak-healthcare"
      pii_model: "models/mom-pii-healthcare"
      threshold: 0.9  # Stricter for healthcare
    
    finance:
      jailbreak_model: "models/mom-jailbreak-finance"
      pii_model: "models/mom-pii-finance"
      threshold: 0.85
    
    technical:
      jailbreak_model: "models/mom-jailbreak-technical"
      threshold: 0.6  # More relaxed for internal tools
      # Falls back to global PII model
    
    # Other domains use global models
    default:
      jailbreak_model: "models/mom-jailbreak-classifier"
      pii_model: "models/mom-pii-classifier"
      threshold: 0.7
```

### Benefits

1. **Higher Accuracy** - Domain-specific models trained on domain-relevant attack patterns
2. **Better Context Understanding** - Models learn domain-specific terminology and patterns
3. **Flexible Security Policies** - Different domains can have different security requirements
4. **Reduced False Positives** - Domain context reduces misclassification
5. **Efficient Resource Usage** - Smaller specialized models can be more efficient than large general models

### Related Files

**Configuration:**
- `config/prompt-guard/jailbreak_domain.yaml` - Per-decision threshold configuration
- `config/prompt-guard/pii_domain.yaml` - Domain-based PII configuration
- `src/semantic-router/pkg/config/registry.go` - Model registry with purpose-based organization

**Training Infrastructure:**
- `src/training/training_lora/prompt_guard_fine_tuning_lora/jailbreak_bert_finetuning_lora.py` - Jailbreak model training
- `src/training/training_lora/pii_model_fine_tuning_lora/pii_bert_finetuning_lora.py` - PII model training
- `src/training/training_lora/classifier_model_fine_tuning_lora/ft_linear_lora.py` - Domain classifier training
- `src/training/training_lora/mmlu_pro_solver_lora/train_all_specialists.sh` - Example of domain-specific training

**Runtime Classification:**
- `src/semantic-router/pkg/classification/classifier.go` - Classifier implementation
- `src/semantic-router/pkg/extproc/req_filter_jailbreak.go` - Jailbreak detection filter
- `src/semantic-router/pkg/services/classification.go` - Classification service

### Implementation Tasks

1. **Training Pipeline** - Extend LoRA training scripts to support domain-specific datasets
2. **Model Registry** - Add domain-specific model entries to `config/registry.go`
3. **Configuration Schema** - Add `domain_guardrails` configuration support
4. **Runtime Selection** - Implement domain-based model selection in classifier
5. **Dataset Preparation** - Create domain-specific guardrail training datasets
6. **Documentation** - Add training guides and configuration examples
7. **Testing** - Add tests for domain-specific guardrail routing

### Related Work

- Domain-specific MMLU solver training: `src/training/training_lora/mmlu_pro_solver_lora/`
- Dual-task training: `src/training/dual_classifier/` (category + PII joint training)
- Per-decision plugin configuration: Already implemented in decision rules

**Comment by satyamg1620** [comment]
@Xunzhuo @rootfs , can you please elaborate more on the `technical` domain part, meaning what it should include?
Also can we use external dataset instead of huggingface for healthcare(jailbreak, pii), finance(jailbreak, pii), technical, etc., for fine-tuning of model

**Comment by rootfs** [comment]
@satyamg1620 Yes, let's collaborate on this one. For immediate delivery, how about we focus on medical and financial sectors? The risk is well documented and there are benchmarks we can use. For example, a recent medical [benchmark](https://www.nature.com/articles/s41746-025-02277-8) on clinical domain is published. A financial benchmark can be found [here](https://aclanthology.org/2025.findings-acl.457.pdf). Although they are not guardrail models, we can still use them to bootstrap. The risk assessment taxonomy should be aligned with [MLCommons](https://arxiv.org/pdf/2503.05731) 

@yossiovadia

**Comment by satyamg1620** [comment]
can we use external dataset for healthcare(jailbreak, pii), finance(jailbreak, pii), technical, etc., Datasets like https://data.mendeley.com/datasets/tzrjx692jy/1 (for Finance PII), https://huggingface.co/datasets/HFXM/CARES-18K (HealthCare jailbreak),https://github.com/verazuo/jailbreak_llms/tree/main (Finance jailbreak), etc. These datasets are also inclined towards PII and jaibreak in specific domain. Let me know if it works otherwise we can work with the benchmark datasets mentioned above.
@rootfs
